### ISSUE TITLE
Lower epoch fences higher: halting process: fenced by envd @ PersistEpoch(Some(2)). ours = PersistEpoch(Some(3))

### ISSUE THREAD

**Issue by def-** [MaterializeInc/materialize#29432]
### What version of Materialize are you using?

cacaf5a403

### What is the issue?

I think this is a regression from https://github.com/MaterializeInc/materialize/pull/29269. It was my understanding that only a higher epoch can fence a lower one: [Txn-wal fencing 2](https://buildkite.com/materialize/nightly/builds/9456#0191c9b1-a453-439b-b32b-31f3b855a237):
```
txn-wal-fencing-mz_second-1  | environmentd: 2024-09-07T00:21:02.632231Z  WARN mz_storage_client::storage_collections: halting process: fenced by envd @ PersistEpoch(Some(2)). ours = PersistEpoch(Some(3))
```
This is not just a bad logging, but the test actually fails:
```
no container found for mz_second-1
==> mzcompose: test case workflow-default failed: materialize.ui.CommandFailureCausedUIError: running docker compose failed (exit status 1)
```
Since Joe and Aljoscha are both out, I'm not sure if a fix or reverting is the best way forward. (potential revert: https://github.com/MaterializeInc/materialize/pull/29433)
```
ci-regexp: halting process: fenced by envd
```

**Comment by nrainer-materialize** [comment]
Further occurrence: https://buildkite.com/materialize/nightly/builds/9490#0191dbb4-8682-4e57-ad3b-6aaa64532596

**Comment by jkosh44** [comment]
From some investigation of the logs and code it looks like the following is happening:

1. `txn-wal-fencing-mz_first-1` starts up with an epoch of 2.
2. `txn-wal-fencing-mz_first-1` sets the opaque value of all storage collections to 2.
3. `txn-wal-fencing-mz_second-1` starts up with an epoch of 3.
4. `txn-wal-fencing-mz_second-1` sets the opaque value of all storage collections to 3.
5. `txn-wal-fencing-mz_second-1` attempts to downgrade the since of some storage collection with a CaS operation. It expects the opaque value to be 3, but it's actually 2, so it panics with `txn-wal-fencing-mz_second-1  | environmentd: 2024-09-07T00:21:02.632231Z  WARN mz_storage_client::storage_collections: halting process: fenced by envd @ PersistEpoch(Some(2)). ours = PersistEpoch(Some(3))`

So the big question is, why is the opaque value 2 and not 3?

**Comment by jkosh44** [comment]
I was able to repro this with some added logs, https://buildkite.com/materialize/nightly/builds/9573#0192010f-e9d4-4192-b652-11cf11be3f6f. We see that the following is indeed happening:

1. `txn-wal-fencing-mz_first-1` creates a materialized view with GID `u48` and sets it's epoch to 2.
```
txn-wal-fencing-mz_first-1   | environmentd: 2024-09-17T17:47:45.260585Z  INFO coord::handle_message{kind="command-execute"}:handle_message:message_command:coord::handle_execute{session="a0276a9f-d312-42f7-8cd4-ba2d425a0c7d"}:coord::handle_execute_inner{stmt="CREATE MATERIALIZED VIEW view23 AS SELECT DISTINCT id, subid, mz_service FROM table23"}:sequence_create_materialized_view:create_materialized_view_finish:coord::catalog_transact_with_side_effects:coord::catalog_transact_with_side_effects::side_effects_fut:storage::create_collections: mz_storage_client::storage_collections: LOOK: Set the epoch of u48 to 2
```

2. `txn-wal-fencing-mz_second-1` starts up and sets the epoch of `u48` to 3.
```
txn-wal-fencing-mz_second-1  | environmentd: 2024-09-17T17:48:50.694186Z  INFO environmentd::run:environmentd::serve:adapter::serve:coord::coordinator:coord::bootstrap:bootstrap_storage_collections:storage::create_collections: mz_storage_client::storage_collections: LOOK: Set the epoch of u48 to 3
```

3. `txn-wal-fencing-mz_first-1` attempts to downgrade the `since` of `u64` and realizes it's been fenced and halts.
```
txn-wal-fencing-mz_first-1   | environmentd: 2024-09-17T17:48:51.621919Z  WARN mz_storage_client::storage_collections: halting process: fenced by envd @ PersistEpoch(Some(3)). ours = PersistEpoch(Some(2)) at id u64
```

4. `txn-wal-fencing-mz_second-1` attempts to downgrade the `since` of `u48`, but the epoch is 2 and it halts.
```
txn-wal-fencing-mz_second-1  | environmentd: 2024-09-17T17:49:21.013942Z  WARN mz_storage_client::storage_collections: halting process: fenced by envd @ PersistEpoch(Some(2)). ours = PersistEpoch(Some(3)) at id u48
```

Once again, the big question is why is the epoch 2 for `u48`? The logs clearly indicate that it was set to 3, and does not show that it was ever set back to 2.

**Comment by jkosh44** [comment]
I'm going to re-open this issue, since I'm actively trying to fix it.

I have a potential explanation. Due to the issue described and fixed in https://github.com/MaterializeInc/materialize/pull/29604, step (2) above, "`txn-wal-fencing-mz_second-1` starts up and sets the epoch of `u48` to 3.", is not guaranteed to actually set the epoch. It's possible that the epoch is not updated, and then later when attempting to downgrade, `u48`, the epoch is still 2, and the new instance panics.

However, this should be an existing issue and there isn't currently an explanation as to why https://github.com/MaterializeInc/materialize/pull/29269 would cause this. Perhaps https://github.com/MaterializeInc/materialize/pull/29269 somehow affects the `since` that is passed in to `compare_and_downgrade_since` and the actual since, which causes us to hit the previously unhit problematic branch.

**Comment by jkosh44** [comment]
I was able to repro with even more logs, https://buildkite.com/materialize/nightly/builds/9584#019201c7-64a1-4284-8504-5dc004bbe103. I can say with certainty that the following is happening:

1. `txn-wal-fencing-mz_first-1` creates a materialized view with GID `u78`, shard ID `s0e5e6590-03c5-4643-8aed-185bd8ed3eb3`, and sets it's epoch to 2.
```
txn-wal-fencing-mz_first-1   | environmentd: 2024-09-17T21:29:29.407224Z  INFO coord::handle_message{kind="command-execute"}:handle_message:message_command:coord::handle_execute{session="c2e1c3fa-8ee1-4493-89eb-30416d027b27"}:coord::handle_execute_inner{stmt="CREATE MATERIALIZED VIEW view38 AS SELECT DISTINCT id, subid, mz_service FROM table38"}:sequence_create_materialized_view:create_materialized_view_finish:coord::catalog_transact_with_side_effects:coord::catalog_transact_with_side_effects::side_effects_fut:storage::create_collections: mz_storage_client::storage_collections: LOOK: Set the epoch of u78 (s0e5e6590-03c5-4643-8aed-185bd8ed3eb3) to 2
```

2. `txn-wal-fencing-mz_second-`1 starts up and _thinks_ it's setting the epoch of `u78` to 3, but actually hits the no-op branch and doesn't change the epoch.
```
txn-wal-fencing-mz_second-1  | environmentd: 2024-09-17T21:30:30.953599Z  INFO environmentd::run:environmentd::serve:adapter::serve:coord::coordinator:coord::bootstrap:bootstrap_storage_collections:storage::create_collections: mz_persist_client::internal::state: LOOK: Oops, I'm not actually updating the opaque value for s0e5e6590-03c5-4643-8aed-185bd8ed3eb3
txn-wal-fencing-mz_second-1  | environmentd: 2024-09-17T21:30:31.019422Z  INFO environmentd::run:environmentd::serve:adapter::serve:coord::coordinator:coord::bootstrap:bootstrap_storage_collections:storage::create_collections: mz_storage_client::storage_collections: LOOK: Set the epoch of u78 (s0e5e6590-03c5-4643-8aed-185bd8ed3eb3) to 3
```
 
3. `txn-wal-fencing-mz_first-1` attempts to downgrade the `since` of `u82` and realizes it's been fenced and halts.
```
txn-wal-fencing-mz_first-1   | environmentd: 2024-09-17T21:30:31.130824Z  WARN mz_storage_client::storage_collections: halting process: fenced by envd @ PersistEpoch(Some(3)). ours = PersistEpoch(Some(2)) at id u82 (s19b6d0ad-2e7b-4aae-9a8c-7d1aaee5de26)
```

4. `txn-wal-fencing-mz_second-1` attempts to downgrade the `since` of `u78`. The expected epoch given to the `compare_and_downgrade_since` call is 3, but the actual epoch is 2, so the call fails and results in a halt.
```
txn-wal-fencing-mz_second-1  | environmentd: 2024-09-17T21:31:28.519529Z  WARN mz_storage_client::storage_collections: halting process: fenced by envd @ PersistEpoch(Some(2)). ours = PersistEpoch(Some(3)) at id u78 (s0e5e6590-03c5-4643-8aed-185bd8ed3eb3)
```

https://github.com/MaterializeInc/materialize/pull/29604 would fully fix this issue. However, the next question is, why is this not happening on `main`?

**Comment by jkosh44** [comment]
I opened up the following PR to try and trigger this failure on `main`: https://github.com/MaterializeInc/materialize/pull/29609/files

The idea was to induce the following ordering:

1. `txn-wal-fencing-mz_second-1` writes a fence token to the durable catalog.
2. `txn-wal-fencing-mz_second-1` opens a `SinceHandle` to some persist shard `s`.
3. `txn-wal-fencing-mz_first-1` downgrades the since of `s`.
4. `txn-wal-fencing-mz_first-1` notices that it has been fenced and halts.
5. `txn-wal-fencing-mz_second-1` attempts to set the epoch of `s` to 3, but hits the problematic branch that doesn't set the epoch because the `since` has been downgraded.
6. `txn-wal-fencing-mz_second-1` attempts to downgrade the since of `s` using an epoch of 3, but fails because the epoch is still at 2.

This ordering was not induced. Below are excerpts of the logs of the txn wal fencing test from the above PR. Some important notes are:
- The `durable catalog open` step is when the fence token is written.
- The `storage collections init` step is when the epoch is set for persist shards.

```
txn-wal-fencing-mz_first-1   | environmentd: 2024-09-17T21:19:42.371302Z  INFO environmentd::run:environmentd::serve: mz_environmentd: startup: envd serve: durable catalog open beginning
txn-wal-fencing-mz_first-1   | environmentd: 2024-09-17T21:19:42.428693Z  INFO environmentd::run:environmentd::serve: mz_environmentd: startup: envd serve: durable catalog open complete in 57.359377ms

txn-wal-fencing-mz_second-1  | environmentd: 2024-09-17T21:20:04.508486Z  INFO environmentd::run:environmentd::serve: mz_environmentd: startup: envd serve: durable catalog open beginning

txn-wal-fencing-mz_first-1   | environmentd: 2024-09-17T21:20:04.555341Z  WARN group_commit_notify:coord::group_commit_initiate{has_write_lock=false}: mz_adapter::util: halting process: unable to confirm leadership: Catalog(Error { kind: Durable(Fence(Epoch { current_epoch: 2, fence_epoch: 3 })) })

txn-wal-fencing-mz_second-1  | environmentd: 2024-09-17T21:20:04.566059Z  INFO environmentd::run:environmentd::serve: mz_environmentd: startup: envd serve: durable catalog open complete in 57.57054ms
txn-wal-fencing-mz_second-1  | environmentd: 2024-09-17T21:20:05.985713Z  INFO environmentd::run:environmentd::serve:adapter::serve:coord::coordinator:coord::bootstrap: mz_adapter::coord: startup: coordinator init: bootstrap: storage collections init beginning

txn-wal-fencing-mz_first-1   | environmentd: 2024-09-17T21:24:04.496841Z  INFO environmentd::run:environmentd::serve: mz_environmentd: startup: envd serve: durable catalog open beginning
txn-wal-fencing-mz_first-1   | environmentd: 2024-09-17T21:24:04.556057Z  INFO environmentd::run:environmentd::serve: mz_environmentd: startup: envd serve: durable catalog open complete in 59.212438ms

txn-wal-fencing-mz_second-1  | environmentd: 2024-09-17T21:24:27.019147Z  INFO environmentd::run:environmentd::serve: mz_environmentd: startup: envd serve: durable catalog open beginning
txn-wal-fencing-mz_second-1  | environmentd: 2024-09-17T21:24:27.083595Z  INFO environmentd::run:environmentd::serve: mz_environmentd: startup: envd serve: durable catalog open complete in 64.445298ms

txn-wal-fencing-mz_first-1   | environmentd: 2024-09-17T21:24:27.102042Z  WARN coord::handle_message{kind="command-commit"}:handle_message:message_command:sequence_end_transaction:coord::group_commit_initiate{has_write_lock=false}: mz_adapter::util: halting process: unable to confirm leadership: Catalog(Error { kind: Durable(Fence(Epoch { current_epoch: 2, fence_epoch: 3 })) })

txn-wal-fencing-mz_second-1  | environmentd: 2024-09-17T21:24:28.503249Z  INFO environmentd::run:environmentd::serve:adapter::serve:coord::coordinator:coord::bootstrap: mz_adapter::coord: startup: coordinator init: bootstrap: storage collections init beginning

txn-wal-fencing-mz_first-1   | environmentd: 2024-09-17T21:28:34.125461Z  INFO environmentd::run:environmentd::serve: mz_environmentd: startup: envd serve: durable catalog open beginning
txn-wal-fencing-mz_first-1   | environmentd: 2024-09-17T21:28:34.207235Z  INFO environmentd::run:environmentd::serve: mz_environmentd: startup: envd serve: durable catalog open complete in 81.771777ms

txn-wal-fencing-mz_second-1  | environmentd: 2024-09-17T21:36:59.119760Z  INFO environmentd::run:environmentd::serve: mz_environmentd: startup: envd serve: durable catalog open beginning

txn-wal-fencing-mz_first-1   | environmentd: 2024-09-17T21:36:59.152139Z  WARN coord::handle_message{kind="command-commit"}:handle_message:message_command:sequence_end_transaction:coord::group_commit_initiate{has_write_lock=false}: mz_adapter::util: halting process: unable to confirm leadership: Catalog(Error { kind: Durable(Fence(Epoch { current_epoch: 2, fence_epoch: 3 })) })

txn-wal-fencing-mz_second-1  | environmentd: 2024-09-17T21:36:59.209572Z  INFO environmentd::run:environmentd::serve: mz_environmentd: startup: envd serve: durable catalog open complete in 89.809546ms
txn-wal-fencing-mz_second-1  | environmentd: 2024-09-17T21:37:00.905434Z  INFO environmentd::run:environmentd::serve:adapter::serve:coord::coordinator:coord::bootstrap: mz_adapter::coord: startup: coordinator init: bootstrap: storage collections init beginning

txn-wal-fencing-mz_first-1   | environmentd: 2024-09-17T21:43:49.911078Z  INFO environmentd::run:environmentd::serve: mz_environmentd: startup: envd serve: durable catalog open beginning
txn-wal-fencing-mz_first-1   | environmentd: 2024-09-17T21:43:49.973745Z  INFO environmentd::run:environmentd::serve: mz_environmentd: startup: envd serve: durable catalog open complete in 62.665156ms

txn-wal-fencing-mz_second-1  | environmentd: 2024-09-17T21:44:12.250092Z  INFO environmentd::run:environmentd::serve: mz_environmentd: startup: envd serve: durable catalog open beginning
txn-wal-fencing-mz_second-1  | environmentd: 2024-09-17T21:44:12.326589Z  INFO environmentd::run:environmentd::serve: mz_environmentd: startup: envd serve: durable catalog open complete in 76.494394ms

txn-wal-fencing-mz_first-1   | environmentd: 2024-09-17T21:44:12.342256Z  WARN coord::handle_message{kind="command-execute"}:handle_message:message_command:coord::handle_execute{session="3ca90f3a-b8cf-4947-80b6-baced45f6006"}:coord::handle_execute_inner{stmt="COMMIT"}:sequence_end_transaction:coord::group_commit_initiate{has_write_lock=false}: mz_adapter::util: halting process: unable to confirm leadership: Catalog(Error { kind: Durable(Fence(Epoch { current_epoch: 2, fence_epoch: 3 })) })

txn-wal-fencing-mz_second-1  | environmentd: 2024-09-17T21:44:13.684053Z  INFO environmentd::run:environmentd::serve:adapter::serve:coord::coordinator:coord::bootstrap: mz_adapter::coord: startup: coordinator init: bootstrap: storage collections init beginning

txn-wal-fencing-mz_first-1   | environmentd: 2024-09-17T21:48:21.505047Z  INFO environmentd::run:environmentd::serve: mz_environmentd: startup: envd serve: durable catalog open beginning
txn-wal-fencing-mz_first-1   | environmentd: 2024-09-17T21:48:21.568053Z  INFO environmentd::run:environmentd::serve: mz_environmentd: startup: envd serve: durable catalog open complete in 63.00248ms

txn-wal-fencing-mz_second-1  | environmentd: 2024-09-17T21:56:36.432636Z  INFO environmentd::run:environmentd::serve: mz_environmentd: startup: envd serve: durable catalog open beginning
txn-wal-fencing-mz_second-1  | environmentd: 2024-09-17T21:56:36.509362Z  INFO environmentd::run:environmentd::serve: mz_environmentd: startup: envd serve: durable catalog open complete in 76.724799ms

txn-wal-fencing-mz_first-1   | environmentd: 2024-09-17T21:56:36.600098Z  WARN group_commit_notify:coord::group_commit_initiate{has_write_lock=false}: mz_adapter::util: halting process: unable to confirm leadership: Catalog(Error { kind: Durable(Fence(Epoch { current_epoch: 2, fence_epoch: 3 })) })

txn-wal-fencing-mz_second-1  | environmentd: 2024-09-17T21:56:38.291056Z  INFO environmentd::run:environmentd::serve:adapter::serve:coord::coordinator:coord::bootstrap: mz_adapter::coord: startup: coordinator init: bootstrap: storage collections init beginning
```

What we see is that `txn-wal-fencing-mz_first-1` is always fenced before `txn-wal-fencing-mz_second-1` starts the `storage collections init`, which prevents the above ordering from happening. In fact `txn-wal-fencing-mz_first-1` is always fenced within one second of `txn-wal-fencing-mz_second-1` writing the fence token. This is expected since timelines are advanced every second, and timeline advancement will confirm leadership. This explains why we never see the issue on `main`.


If we take a look at the same logs, but from a failed run from the PR that caused this issue (https://github.com/MaterializeInc/materialize/pull/29593), we see the following (Note panics don't have timestamps, so I've included the log lines immediately before and after each panic):

```
txn-wal-fencing-mz_first-1  | environmentd: 2024-09-17T21:20:14.829793Z  INFO environmentd::run:environmentd::serve: mz_environmentd: startup: envd serve: durable catalog open beginning
txn-wal-fencing-mz_first-1  | environmentd: 2024-09-17T21:20:14.892546Z  INFO environmentd::run:environmentd::serve: mz_environmentd: startup: envd serve: durable catalog open complete in 62.750269ms

txn-wal-fencing-mz_second-1  | environmentd: 2024-09-17T21:20:28.098127Z  INFO environmentd::run:environmentd::serve: mz_environmentd: startup: envd serve: durable catalog open beginning
txn-wal-fencing-mz_second-1  | environmentd: 2024-09-17T21:20:28.153227Z  INFO environmentd::run:environmentd::serve: mz_environmentd: startup: envd serve: durable catalog open complete in 55.098214ms
txn-wal-fencing-mz_second-1  | environmentd: 2024-09-17T21:20:29.556982Z  INFO environmentd::run:environmentd::serve:adapter::serve:coord::coordinator:coord::bootstrap: mz_adapter::coord: startup: coordinator init: bootstrap: storage collections init beginning

txn-wal-fencing-mz_first-1   | environmentd: 2024-09-17T21:20:38.661627Z  INFO mz_storage_controller::collection_mgmt: upper mismatch while bumping upper, syncing to persist state self.id=s674 actual_upper=1726608037979 expected_upper=1726608036979
txn-wal-fencing-mz_first-1   | thread 'coordinator' panicked at src/adapter/src/util.rs:287:23:
txn-wal-fencing-mz_first-1   | unable to confirm leadership: Catalog(Error { kind: Durable(Fence(Epoch { current_epoch: 2, fence_epoch: 3 })) })
txn-wal-fencing-mz_second-1  | environmentd: 2024-09-17T21:20:38.955591Z  INFO mz_storage_controller::collection_mgmt: upper mismatch while bumping upper, syncing to persist state self.id=s675 actual_upper=1726608038292 expected_upper=1726608037291

txn-wal-fencing-mz_first-1   | environmentd: 2024-09-17T21:24:43.635683Z  INFO environmentd::run:environmentd::serve: mz_environmentd: startup: envd serve: durable catalog open beginning
txn-wal-fencing-mz_first-1   | environmentd: 2024-09-17T21:24:43.692224Z  INFO environmentd::run:environmentd::serve: mz_environmentd: startup: envd serve: durable catalog open complete in 56.538511ms

txn-wal-fencing-mz_second-1  | environmentd: 2024-09-17T21:24:56.542740Z  INFO environmentd::run:environmentd::serve: mz_environmentd: startup: envd serve: durable catalog open beginning
txn-wal-fencing-mz_second-1  | environmentd: 2024-09-17T21:24:56.596409Z  INFO environmentd::run:environmentd::serve: mz_environmentd: startup: envd serve: durable catalog open complete in 53.664796ms
txn-wal-fencing-mz_second-1  | environmentd: 2024-09-17T21:24:58.009522Z  INFO environmentd::run:environmentd::serve:adapter::serve:coord::coordinator:coord::bootstrap: mz_adapter::coord: startup: coordinator init: bootstrap: storage collections init beginning

txn-wal-fencing-mz_first-1   | environmentd: 2024-09-17T21:25:07.520799Z  INFO mz_storage_controller::collection_mgmt: upper mismatch while bumping upper, syncing to persist state self.id=s674 actual_upper=1726608307473 expected_upper=1726608306474
txn-wal-fencing-mz_first-1   | thread 'coordinator' panicked at src/adapter/src/util.rs:287:23:
txn-wal-fencing-mz_first-1   | unable to confirm leadership: Catalog(Error { kind: Durable(Fence(Epoch { current_epoch: 2, fence_epoch: 3 })) })
txn-wal-fencing-mz_first-1   | cluster-s2-replica-s2-gen-0: 2024-09-17T21:25:07.927099Z  WARN mz_persist_client::rpc: pubsub client error: Status { code: Unknown, message: "h2 protocol error: error reading a body from connection", source: Some(hyper::Error(Body, Error { kind: Io(Custom { kind: BrokenPipe, error: "stream closed because of a broken pipe" }) })) }

txn-wal-fencing-mz_first-1   | environmentd: 2024-09-17T21:29:02.353683Z  INFO environmentd::run:environmentd::serve: mz_environmentd: startup: envd serve: durable catalog open beginning
txn-wal-fencing-mz_first-1   | environmentd: 2024-09-17T21:29:02.413048Z  INFO environmentd::run:environmentd::serve: mz_environmentd: startup: envd serve: durable catalog open complete in 59.361185ms

txn-wal-fencing-mz_second-1  | environmentd: 2024-09-17T21:30:27.947980Z  INFO environmentd::run:environmentd::serve: mz_environmentd: startup: envd serve: durable catalog open beginning
txn-wal-fencing-mz_second-1  | environmentd: 2024-09-17T21:30:28.042133Z  INFO environmentd::run:environmentd::serve: mz_environmentd: startup: envd serve: durable catalog open complete in 94.149823ms
txn-wal-fencing-mz_second-1  | environmentd: 2024-09-17T21:30:29.911555Z  INFO environmentd::run:environmentd::serve:adapter::serve:coord::coordinator:coord::bootstrap: mz_adapter::coord: startup: coordinator init: bootstrap: storage collections init beginning

txn-wal-fencing-mz_first-1   | environmentd: 2024-09-17T21:30:31.130824Z  WARN mz_storage_client::storage_collections: halting process: fenced by envd @ PersistEpoch(Some(3)). ours = PersistEpoch(Some(2)) at id u82 (s19b6d0ad-2e7b-4aae-9a8c-7d1aaee5de26)

txn-wal-fencing-mz_second-1  | environmentd: 2024-09-17T21:31:28.519529Z  WARN mz_storage_client::storage_collections: halting process: fenced by envd @ PersistEpoch(Some(2)). ours = PersistEpoch(Some(3)) at id u78 (s0e5e6590-03c5-4643-8aed-185bd8ed3eb3)
``` 

Here we notice the following two things:

- It takes longer than 1 second for `txn-wal-fencing-mz_first-1` to realize it has been fenced. In fact in one instance it took 10 seconds.
- The problematic ordering described above is possible. Specifically the `storage collections init` step for `txn-wal-fencing-mz_second-1` starts before `txn-wal-fencing-mz_first-1` notices that it has been fenced.


The next question is, why does it take so longer for a fence to be noticed in my branch?

**Comment by jkosh44** [comment]
> The next question is, why does it take so longer for a fence to be noticed in my branch?

I've added some more logs, and it turns out that on my branch the old instance notices within a second that it's been fenced, but it takes awhile (sometimes up to 10 seconds) for the process to shut down. The reason is that we changed the fence error from a `halt!` to a `panic!`. Panics can be much slower and are known to take multiple seconds in Materialize. In fact, if we change the `panic!` to an `exit`, then the issue goes away entirely.

So in conclusion. The original PR changed the timing of fencing, which triggered an existing issue, which has now been fixed by https://github.com/MaterializeInc/materialize/pull/29604.
